### ISSUE TITLE
[Enhancement] Add the ability to filter items in Association Field

### ISSUE THREAD

**Issue by bzerangue** [symphonists/association_field#12]
It would be great to add the ability to filter items in Association Field like the old Subsection Manager did...

![](http://f.cl.ly/items/2K2v0r38071Y0B3P3i3n/Screen%20Shot%202014-10-08%20at%204.14.50%20PM.png)

That would be helpful if you don't want all the results showing up from the section that you are associating with.

**Comment by bzerangue** [comment]
Any thoughts if filtering would be added?

**Comment by jonmifsud** [comment]
I will most certainly need to add this functionality for a client, though I didn't have time to dive into it yet. I'm hoping to have a look into it within the next couple of weeks.

**Comment by jurajkapsz** [comment]
Guys, what are your use cases for filtered options? Just wanted to check on mine (mentioned bellow), or get some more possible use cases.
And does it also means, that you would add tags as additional data to your entries for this filtering feature?

I was using SSM with filtering on a section holding lets say "content sources" like named groups of images, documents etc. Now I had to broke it to separate sections, but maybe its a bit better like this.

**Comment by jonmifsud** [comment]
My main use case is the following:

The client has a mixture of content types which are very similar - eg
'articles' and 'blog articles' they more or less have the same
relationships to categories/news sections. Functionality wise it would be
an overkill to duplicate all the fields and feature requests considering
all the fields are otherwise exactly the same.

My idea would be to apply filtering like what you would see in the backend
(have to confirm support) and ideally I want to use javascript data
attributes so I can switch parameters on the fly. This would give maximum
flexibility but might be out of scope for the association field itself.

**Comment by jurajkapsz** [comment]
Thanks for the example @jonmifsud . So you are having a single section for all the articles. The filtering idea is pretty advanced for me, but let's see what's possible first.

**Comment by jonmifsud** [comment]
Yes more or less. I have to work on an import/data migration next week.
I'll be switching to filtering once I get that all working, so I presume it
would be around a couple of weeks till I get to it. Obviously if someone
has some ideas for that I'm pretty open once I'll be working on it.

**Comment by nilshoerrmann** [comment]
If you'd like to implement this, I think the following steps are needed:

1. A new setting needs to be added to the field (database and UI). In the UI, SSM used a tag list to suggest all tag list or select list values of the associated sections.
2. The [`findOptions()`](https://github.com/symphonists/association_field/blob/master/fields/field.association.php#L89-L144) function needs to take this new setting into account. If you follow the logic of SSM, you'd either have no filtering, matches (`articles`), exclusions (`-comments`) or a mixture of the latter (`articles, -comments`).
3. If you'd like the Association UI Selector to take that new field setting into account, the [`query` endpoint](https://github.com/hananils/association_ui_selector/blob/master/content/content.query.php) for the search needs to filter the output as well.

**Comment by jonmifsud** [comment]
@nilshoerrmann, Thanks for the pointers, I decided to start off with work on the Association UI Selector, I've modified the Query Endpoint and the `publish.js` script. So it can now accept url param filters (similar to the ones supported by the backend filters). And that seems to work great, I think the filters should be far more flexible the the old SSM.

If you think that does the trick I will be adding the Filtering UI to the association field itself. I was thinking it would be nice to go with just the Association UI Selector but `setAssociationContext` is part of the core, and not sure I should be messing with that.

Btw in which context is `findOptions()` used? is it only for the pre-filled textbox? As the Associations UI seems to do a fetch with Ajax straight away when the page is loaded.

**Comment by nilshoerrmann** [comment]
> Btw in which context is findOptions() used?

Yes, it's used for the basic select box used if no interface is selected.
